### PR TITLE
[pytest] support for integrating platform.json static data retrieval for platform_api tests

### PR DESCRIPTION
**Pull request by vdahiya12**
This PR adds the  support for incorporating
platform.json as a de-facto gold standard for
comparing the values returned by platform_api.
The enhancement addresses adding the
chassis and interfaces dictionary within the duthost-facts
derived from platform.json file  which in turn would be utilized
to carry out the comparison for all components of platform tests 

Signed-off-by: vaibhav-dahiya <vdahiya@microsoft.com>

### CODE REVIEW

File: tests/common/devices.py [hunk 1]
```diff
@@ -268,6 +268,19 @@ def _get_platform_info(self):
...
+            platform_file_path = os.path.join('/usr/share/sonic/device',result["platform"],'platform.json')
+
+            try:
+                out = self.command("cat {}".format(platform_file_path))
+                platform_info = json.loads(out["stdout"])
+                result["chassis"] = platform_info["chassis"]
+            except:
+                # if platform.json does not exist, then it's not added currently for certain platforms
+                # eventually all the platforms should have the platform.json
+                pass
```
Comment (wangxin) [inline tests/common/devices.py:282]: For easier troubleshooting, tt would be better to log something here in case an exception is caught.

File: tests/platform_tests/api/test_fan_drawer.py [hunk 2]
```diff
@@ -40,6 +41,11 @@ def setup(self, platform_api_conn):
             except:
                 pytest.fail("num_fan_drawers is not an integer")
 
+        chassis = duthost.facts['chassis']
```
Comment (wangxin) [inline tests/platform_tests/api/test_fan_drawer.py:44]: In case there is no platform.json file on DUT or something went wrong with reading and parsing the platform.json file, accessing `duthost.facts['chassis']` may trigger KeyError exception. It's better to use:
```
chassis = duthost.facts.get('chassis', None)
```

File: tests/common/devices.py [hunk 3]
```diff
@@ -268,6 +268,19 @@ def _get_platform_info(self):
                 result["hwsku"] = line.split(":")[1].strip()
             elif line.startswith("ASIC:"):
                 result["asic_type"] = line.split(":")[1].strip()
+
+        if result["platform"]:
+            platform_file_path = os.path.join('/usr/share/sonic/device',result["platform"],'platform.json')
```
Comment (jleveque) [inline tests/common/devices.py:273]: Add spaces after commas.

**Review by jleveque — CHANGES_REQUESTED** [review]

File: tests/common/devices.py (hunk 1 above)
Comment (vdahiya12) [reply tests/common/devices.py:284]: fixed

File: tests/platform_tests/api/test_fan_drawer.py (hunk 2 above)
Comment (vdahiya12) [reply tests/platform_tests/api/test_fan_drawer.py]: fixed

File: tests/common/devices.py (hunk 3 above)
Comment (vdahiya12) [reply tests/common/devices.py]: fixed

File: tests/common/devices.py [hunk 4]
```diff
@@ -268,6 +268,21 @@ def _get_platform_info(self):
...
             elif line.startswith("ASIC:"):
                 result["asic_type"] = line.split(":")[1].strip()
+
+        if result["platform"]:
+            platform_file_path = os.path.join('/usr/share/sonic/device', result["platform"], 'platform.json')
+
+            try:
+                out = self.command("cat {}".format(platform_file_path))
+                platform_info = json.loads(out["stdout"])
+                result["chassis"] = platform_info["chassis"]
```
Comment (jleveque) [inline tests/common/devices.py:278]: Rather than explicitly adding `"chassis"` to `result` here, can you add all the elements parsed from the file? This way the code will not need to be modified if/when more elements are added to the file.

File: tests/common/devices.py [hunk 5]
```diff
@@ -268,6 +268,21 @@ def _get_platform_info(self):
...
+            except:
+                # if platform.json does not exist, then it's not added currently for certain platforms
+                # eventually all the platforms should have the platform.json
+                logging.debug("platform.json is not available for this platform, dut facts will not contain
+                              chassis and interfaces data")
```
Comment (jleveque) [inline tests/common/devices.py:283]: As with the above comment, this message should be more generic, and not specify "chassis and interfaces data". Suggest rewording along the lines of,

```
"platform.json is not available for this platform. DUT facts will not contain complete platform information."
```

**Review by jleveque — CHANGES_REQUESTED** [review]

**Comment by daall** [comment]
> This pull request **introduces 3 alerts** when merging [11104ff](https://github.com/Azure/sonic-mgmt/commit/11104ff5ff88ba07b8ececbb62483866a381d59d) into [2cd7618](https://github.com/Azure/sonic-mgmt/commit/2cd7618f9263f91fc6fa8a9f03403e9266938df5) - [view on LGTM.com](https://lgtm.com/projects/g/Azure/sonic-mgmt/rev/pr-b5250258c52affeeb1c396642cab6c3d233bbf31)
> 
> **new alerts:**
> 
> * 1 for Unused local variable
> * 1 for Except block handles 'BaseException'
> * 1 for Unused import

Can you fix these as well?

File: tests/platform_tests/api/test_fan_drawer.py [hunk 6]
```diff
@@ -40,6 +41,11 @@ def setup(self, platform_api_conn):
             except:
                 pytest.fail("num_fan_drawers is not an integer")
 
+        chassis = duthost.facts.get('chassis', None)
+
+        if chassis and len(chassis['fan_drawers']) > 0 :
```
Comment (daall) [inline tests/platform_tests/api/test_fan_drawer.py:46]:
```suggestion
        if chassis and len(chassis['fan_drawers']) > 0:
```

File: tests/platform_tests/api/test_fan_drawer.py [hunk 7]
```diff
@@ -49,6 +55,8 @@ def test_get_name(self, duthost, localhost, platform_api_conn):
 
             if self.expect(name is not None, "Unable to retrieve Fan_drawer {} name".format(i)):
                 self.expect(isinstance(name, str), "Fan_drawer {} name appears incorrect".format(i))
+                if fan_drawer_truth:
+                    self.expect(name == fan_drawer_truth[i]['name'], "Fan_drawer {} name does not match".format(i))
```
Comment (daall) [inline tests/platform_tests/api/test_fan_drawer.py:59]: Can we include the expected name as well? It might be more clear where the mismatch is if we can compare expected v. received.

File: tests/platform_tests/api/test_fan_drawer.py [hunk 8]
```diff
@@ -98,6 +106,8 @@ def test_get_num_fans(self, duthost, localhost, platform_api_conn):
             num_fans = fan_drawer.get_num_fans(platform_api_conn, i)
             if self.expect(num_fans is not None, "Unable to retrieve fan_drawer {} number of fans".format(i)):
                 self.expect(isinstance(num_fans, int), "fan drawer {} number of fans appear to be incorrect".format(i))
+                if fan_drawer_truth:
+                    self.expect(num_fans == fan_drawer_truth[i]['num_fans'], "Fan_drawer {} number of fans does not match".format(i))
```
Comment (daall) [inline tests/platform_tests/api/test_fan_drawer.py:110]: Same feedback as above.

**Review by daall — CHANGES_REQUESTED** [review]
See comments + LGTM warnings.

File: tests/platform_tests/api/test_fan_drawer.py (hunk 6 above)
Comment (vdahiya12) [reply tests/platform_tests/api/test_fan_drawer.py]: fixed

File: tests/platform_tests/api/test_fan_drawer.py (hunk 7 above)
Comment (vdahiya12) [reply tests/platform_tests/api/test_fan_drawer.py]: fixed

File: tests/platform_tests/api/test_fan_drawer.py (hunk 8 above)
Comment (vdahiya12) [reply tests/platform_tests/api/test_fan_drawer.py]: fixed

File: tests/common/devices.py (hunk 5 above)
Comment (vdahiya12) [reply tests/common/devices.py]: fixed

File: tests/common/devices.py (hunk 4 above)
Comment (vdahiya12) [reply tests/common/devices.py]: fixed

File: tests/platform_tests/api/test_fan_drawer.py [hunk 9]
```diff
@@ -40,6 +41,10 @@ def setup(self, platform_api_conn):
             except:
```
Comment (jleveque) [inline tests/platform_tests/api/test_fan_drawer.py:41]: LGTM is complaining about this. Change to `except Exception:`

**Review by jleveque — CHANGES_REQUESTED** [review]

File: tests/platform_tests/api/test_fan_drawer.py (hunk 9 above)
Comment (vdahiya12) [reply tests/platform_tests/api/test_fan_drawer.py:41]: changed as suggested, fixed

File: tests/common/devices.py [hunk 10]
```diff
@@ -268,6 +268,22 @@ def _get_platform_info(self):
...
                 result["asic_type"] = line.split(":")[1].strip()
+
+        if result["platform"]:
+            platform_file_path = os.path.join('/usr/share/sonic/device', result["platform"], 'platform.json')
+
+            try:
+                out = self.command("cat {}".format(platform_file_path))
+                platform_info = json.loads(out["stdout"])
+                result["chassis"] = platform_info.get('chassis', None)
+                result["interfaces"] = platform_info.get('interfaces', None)
```
Comment (jleveque) [inline tests/common/devices.py:279]: Can we eliminate the need to know what keys are available? Optimally, we would like to insert all data from platform.json into `result`. Can you make this more generic?

**Review by jleveque — CHANGES_REQUESTED** [review]

File: tests/common/devices.py [hunk 11]
```diff
@@ -268,6 +268,22 @@ def _get_platform_info(self):
...
+
+             except Exception:
```
Comment (jleveque) [inline tests/common/devices.py:281]: Indentation now has an extra space, causing LGTM to report a syntax error.

**Review by jleveque — CHANGES_REQUESTED** [review]

File: tests/common/devices.py (hunk 10 above)
Comment (vdahiya12) [reply tests/common/devices.py]: Thanks, converted this to a for loop.
Should be fixed

File: tests/common/devices.py (hunk 11 above)
Comment (vdahiya12) [reply tests/common/devices.py]: Thanks, 
Fixed

File: tests/platform_tests/api/test_fan_drawer.py [hunk 12]
```diff
@@ -49,6 +54,8 @@ def test_get_name(self, duthost, localhost, platform_api_conn):
 
             if self.expect(name is not None, "Unable to retrieve Fan_drawer {} name".format(i)):
                 self.expect(isinstance(name, str), "Fan_drawer {} name appears incorrect".format(i))
+                if self.fan_drawer_truth:
+                    self.expect(name == self.fan_drawer_truth[i]['name'], "Fan_drawer {} name does not match,  expected name {} ".format(i, self.fan_drawer_truth[i]['name']))
```
Comment (jleveque) [inline tests/platform_tests/api/test_fan_drawer.py:58]: Remove extra space before "expected"

File: tests/platform_tests/api/test_fan_drawer.py [hunk 13]
```diff
@@ -98,6 +105,8 @@ def test_get_num_fans(self, duthost, localhost, platform_api_conn):
             num_fans = fan_drawer.get_num_fans(platform_api_conn, i)
             if self.expect(num_fans is not None, "Unable to retrieve fan_drawer {} number of fans".format(i)):
                 self.expect(isinstance(num_fans, int), "fan drawer {} number of fans appear to be incorrect".format(i))
+                if self.fan_drawer_truth:
+                    self.expect(name == self.fan_drawer_truth[i]['num_fans'], "Fan_drawer {} num_fans does not match,  expected num_fans {} ".format(i, self.fan_drawer_truth[i]['num_fans']))
```
Comment (jleveque) [inline tests/platform_tests/api/test_fan_drawer.py:109]: Remove extra space before "expected"

**Review by jleveque — CHANGES_REQUESTED** [review]

File: tests/platform_tests/api/test_fan_drawer.py (hunk 12 above)
Comment (vdahiya12) [reply tests/platform_tests/api/test_fan_drawer.py]: fixed

File: tests/platform_tests/api/test_fan_drawer.py (hunk 13 above)
Comment (vdahiya12) [reply tests/platform_tests/api/test_fan_drawer.py]: fixed

File: tests/common/devices.py [hunk 14]
```diff
@@ -268,6 +268,22 @@ def _get_platform_info(self):
...
+                out = self.command("cat {}".format(platform_file_path))
+                platform_info = json.loads(out["stdout"])
+                for key, value in platform_info.iteritems():
+                    result[key] = value
+
+            except Exception:
+                # if platform.json does not exist, then it's not added currently for certain platforms
+                # eventually all the platforms should have the platform.json
+                logging.debug("platform.json is not available for this platform."
+                              + "DUT facts will not contain complete platform information.")
```
Comment (jleveque) [inline tests/common/devices.py:285]: Looks like these strings will be concatenated without a space in between.

**Review by jleveque — CHANGES_REQUESTED** [review]

**Review by jleveque — APPROVED** [review]
LGTM. @daall: Can you review again, please?

File: tests/common/devices.py (hunk 14 above)
Comment (vdahiya12) [reply tests/common/devices.py]: added a space, fixed

File: tests/common/devices.py [hunk 15]
```diff
@@ -268,6 +268,22 @@ def _get_platform_info(self):
                 result["hwsku"] = line.split(":")[1].strip()
             elif line.startswith("ASIC:"):
                 result["asic_type"] = line.split(":")[1].strip()
+
+        if result["platform"]:
+            platform_file_path = os.path.join('/usr/share/sonic/device', result["platform"], 'platform.json')
```
Comment (daall) [inline tests/common/devices.py:273]: There's a mix of single and double quotes throughout, can you make them all double quotes for consistency? (unless it's a character or string inside a string or something)

**Review by daall — APPROVED** [review]

File: tests/platform_tests/api/test_fan_drawer.py [hunk 16]
```diff
@@ -24,22 +24,27 @@
...
+    def setup(self, duthost, platform_api_conn):
         if self.num_fan_drawers is None:
             try:
                 self.num_fan_drawers = int(chassis.get_num_fan_drawers(platform_api_conn))
             except:
                 pytest.fail("num_fan_drawers is not an integer")
 
+        chassis_truth = duthost.facts.get('chassis', None)
+        if chassis_truth:
+            self.fan_drawer_truth = chassis_truth.get('fan_drawers', None)
```
Comment (jleveque) [inline tests/platform_tests/api/test_fan_drawer.py:46]: I think there should be an `else` clause here which logs a warning message stating that the no platform facts were found, so the test will not be as thorough.

**Review by jleveque — CHANGES_REQUESTED** [review]

File: tests/platform_tests/api/test_fan_drawer.py [hunk 17]
```diff
@@ -24,22 +24,29 @@
...
```
Comment (jleveque) [inline tests/platform_tests/api/test_fan_drawer.py:46]: You should probably also confirm that `self.fan_drawer_truth` is not None and log a similar message.

**Review by jleveque — CHANGES_REQUESTED** [review]

**Review by jleveque — APPROVED** [review]

**Comment by dipali0125** [comment]
@jleveque  Is there a HLD (definition) for this platform.json file.

**Comment by jleveque** [comment]
@dipali0125: There is not HLD, but a section has been added to the [SONiC Porting Guide](https://github.com/Azure/SONiC/wiki/Porting-Guide#common-device-specific-file-details). More details will be added to this section in the near future.
